File: apps/desktop/src/lib/api/links.ts
/**
 * src/lib/api/links.ts — Tauri IPC wrappers for entity links.
 *
 * Links are bookmarks. 900CRM stores the URL or file path as text and does
 * not copy or upload the file.
 */

import { invoke } from '@tauri-apps/api/core';
import type { CrmEntityType } from './notes';

export type LinkEntityType = Exclude<CrmEntityType, 'activity'>;
export type EntityLinkKind = 'url' | 'path';

export interface EntityLink {
  id: string;
  entityType: LinkEntityType;
  entityId: string;
  title: string;
  kind: EntityLinkKind;
  target: string;
  createdAt: string;
  updatedAt: string;
  deletedAt: string | null;
  deviceId: string;
}

export interface CreateEntityLinkPayload {
  entityType: LinkEntityType;
  entityId: string;
  title?: string | null;
  kind: EntityLinkKind;
  target: string;
}

interface BackendEntityLink {
  id: string;
  entity_type: LinkEntityType;
  entity_id: string;
  title: string;
  kind: string;
  target: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  device_id: string;
}

function mapLink(link: BackendEntityLink): EntityLink {
  return {
    id: link.id,
    entityType: link.entity_type,
    entityId: link.entity_id,
    title: link.title,
    kind: link.kind === 'path' ? 'path' : 'url',
    target: link.target,
    createdAt: link.created_at,
    updatedAt: link.updated_at,
    deletedAt: link.deleted_at ?? null,
    deviceId: link.device_id,
  };
}

function normalizeText(value: string): string {
  return value.trim();
}

export async function createEntityLink(data: CreateEntityLinkPayload): Promise<EntityLink> {
  const link = await invoke<BackendEntityLink>('create_entity_link', {
    entity_type: data.entityType,
    entity_id: normalizeText(data.entityId),
    title: data.title?.trim() || null,
    kind: data.kind,
    target: normalizeText(data.target),
  });
  return mapLink(link);
}

export async function listEntityLinks(
  entityType: LinkEntityType,
  entityId: string,
): Promise<EntityLink[]> {
  const links = await invoke<BackendEntityLink[]>('list_entity_links', {
    entity_type: entityType,
    entity_id: normalizeText(entityId),
  });
  return links.map(mapLink);
}

export async function updateEntityLink(
  id: string,
  data: Pick<CreateEntityLinkPayload, 'title' | 'kind' | 'target'>,
): Promise<EntityLink> {
  const link = await invoke<BackendEntityLink>('update_entity_link', {
    id: normalizeText(id),
    title: data.title?.trim() || null,
    kind: data.kind,
    target: normalizeText(data.target),
  });
  return mapLink(link);
}

export async function deleteEntityLink(id: string): Promise<void> {
  await invoke<void>('delete_entity_link', { id: normalizeText(id) });
}

File: crates/crm-core/src/storage/links.rs
//! Entity link CRUD for 900CRM.
//!
//! Links are bookmarks attached to a contact, organization, or deal. They store
//! either an http(s) URL or a local file path as text. 900CRM does not copy or
//! upload files.

use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};

use crate::utils::{
    datetime::now_iso8601,
    errors::{CrmError, CrmResult},
    uuid::new_uuid,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityLink {
    pub id: String,
    pub entity_type: String,
    pub entity_id: String,
    pub title: String,
    pub kind: String,
    pub target: String,
    pub created_at: String,
    pub updated_at: String,
    pub deleted_at: Option<String>,
    pub device_id: String,
}

pub fn create_link(
    conn: &Connection,
    entity_type: &str,
    entity_id: &str,
    title: &str,
    kind: &str,
    target: &str,
    device_id: &str,
) -> CrmResult<EntityLink> {
    let id = new_uuid();
    let now = now_iso8601();

    conn.execute(
        r#"
        INSERT INTO entity_links
            (id, entity_type, entity_id, title, kind, target, created_at, updated_at, device_id)
        VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)
        "#,
        params![
            id,
            entity_type,
            entity_id,
            title,
            kind,
            target,
            now,
            now,
            device_id
        ],
    )?;

    get_link(conn, &id)
}

pub fn get_link(conn: &Connection, id: &str) -> CrmResult<EntityLink> {
    conn.query_row(
        r#"
        SELECT id, entity_type, entity_id, title, kind, target,
               created_at, updated_at, deleted_at, device_id
        FROM entity_links
        WHERE id = ?1 AND deleted_at IS NULL
        "#,
        params![id],
        row_to_link,
    )
    .map_err(|e| match e {
        rusqlite::Error::QueryReturnedNoRows => {
            CrmError::NotFound(format!("Link '{}' not found", id))
        }
        other => CrmError::Database(other.to_string()),
    })
}

pub fn list_links_for_entity(
    conn: &Connection,
    entity_type: &str,
    entity_id: &str,
) -> CrmResult<Vec<EntityLink>> {
    let mut stmt = conn.prepare(
        r#"
        SELECT id, entity_type, entity_id, title, kind, target,
               created_at, updated_at, deleted_at, device_id
        FROM entity_links
        WHERE entity_type = ?1 AND entity_id = ?2 AND deleted_at IS NULL
        ORDER BY created_at DESC, id DESC
        "#,
    )?;

    let rows = stmt.query_map(params![entity_type, entity_id], row_to_link)?;
    Ok(rows.filter_map(|row| row.ok()).collect())
}

pub fn update_link(
    conn: &Connection,
    id: &str,
    title: &str,
    kind: &str,
    target: &str,
) -> CrmResult<EntityLink> {
    let now = now_iso8601();
    let changed = conn.execute(
        r#"
        UPDATE entity_links
        SET title = ?1, kind = ?2, target = ?3, updated_at = ?4
        WHERE id = ?5 AND deleted_at IS NULL
        "#,
        params![title, kind, target, now, id],
    )?;

    if changed == 0 {
        return Err(CrmError::NotFound(format!("Link '{}' not found", id)));
    }

    get_link(conn, id)
}

pub fn soft_delete_link(conn: &Connection, id: &str) -> CrmResult<()> {
    let now = now_iso8601();
    let changed = conn.execute(
        "UPDATE entity_links SET deleted_at = ?1, updated_at = ?1 WHERE id = ?2 AND deleted_at IS NULL",
        params![now, id],
    )?;

    if changed == 0 {
        return Err(CrmError::NotFound(format!(
            "Link '{}' not found or already deleted",
            id
        )));
    }

    Ok(())
}

fn row_to_link(row: &rusqlite::Row<'_>) -> rusqlite::Result<EntityLink> {
    Ok(EntityLink {
        id: row.get(0)?,
        entity_type: row.get(1)?,
        entity_id: row.get(2)?,
        title: row.get(3)?,
        kind: row.get(4)?,
        target: row.get(5)?,
        created_at: row.get(6)?,
        updated_at: row.get(7)?,
        deleted_at: row.get(8)?,
        device_id: row.get(9)?,
    })
}
